### ISSUE TITLE
DataFrame Renderer

### ISSUE THREAD

**Issue by EronWright** [andypetrella/spark-notebook#200]
The DataFrame (formerly known as SchemaRDD) is a new core abstraction in Spark.    Major subsystems such as Spark SQL and the new Spark ML are based on DataFrames.   DataFrames are column-oriented, schematized tables.

A table-like renderer should be provided, resembling the output from DataFrame::show().  Ideally the table would be paged and capable of lazily exploring the entire contents of the DataFrame.

**Comment by andypetrella** [comment]
:+1:  having both the schema and the paginable table in two different tabs for instance?

**Comment by EronWright** [comment]
A few words on the data types.  DataFrame supports a wide variety of column types, defined here:
https://github.com/apache/spark/tree/branch-1.4/sql/catalyst/src/main/scala/org/apache/spark/sql/types

Note that UDTs also exist, such as [VectorUDT](https://github.com/apache/spark/blob/branch-1.4/mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala#L164).  

The DataType classes facilitate language-neutral serialization of values, both for caching purposes and for python support.

Another curiosity is column metadata or *attributes*.  For example, the [StringIndexer](https://github.com/apache/spark/blob/branch-1.4/mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala) transforms a column of string labels into a numerical index, with the lookup table attached as a column attribute.   Makes for useful hints to a UI.  Further information in the [SPARK-4588 design doc](https://docs.google.com/document/d/1796XfSzFbZvGWFs0ky99AJhlqkOBRG1O2bUxK2N4Grk/edit).

**Comment by EronWright** [comment]
Feeling ambitious?  Another feature here would be to support custom cell renderers.    Maybe the renderer would be selected based on the UDT or based on column attributes.

**Comment by andypetrella** [comment]
@EronWright cool references indeed. Very helpful when we'll start the work.

I think there is already some stuffs in the notebook that can be reused for this one.
Best is to move with small steps to have something usable quickly and then iterate on it.

**what exists**
The [Reflector](https://github.com/andypetrella/spark-notebook/blob/master/modules/common/src/main/scala/notebook/util/Reflector.scala) is able to take a type, deconstruct it (first level only, the heuristic is trivial atm) and returns a mapping `name` → `value`.

The reflector will be used in [widgets](https://github.com/andypetrella/spark-notebook/blob/master/modules/common/src/main/scala/notebook/front/widgets/package.scala#L253) to convert a list of objects to a list of maps (using the typeclass `MagicRenderPoint`).

This conversion will allow to things:
* conversion to Json to be used on the client side, see [here](https://github.com/andypetrella/spark-notebook/blob/master/modules/common/src/main/scala/notebook/front/widgets/package.scala#L213) and [there](https://github.com/andypetrella/spark-notebook/blob/master/modules/common/src/main/scala/notebook/front/widgets/package.scala#L282)
* define defaults charts based on the components: see [here](https://github.com/andypetrella/spark-notebook/blob/master/modules/common/src/main/scala/notebook/front/widgets/package.scala#L442)

That means that we could have something quick for both charts/table of `DataFrame` and auto plotting by having a conversion more generic than only for `Seq`, `Array` or `Map`:
* [here](https://github.com/andypetrella/spark-notebook/blob/master/modules/common/src/main/scala/notebook/front/widgets/package.scala#L258) is the automatic conversions to the magic points that adds the metadata necessary for the auto plotting
* [there](https://github.com/andypetrella/spark-notebook/blob/master/modules/common/src/main/scala/notebook/front/Renderer.scala#L52) is for the auto plotting based on the container type (need to add DataFrame)

---

[not tested]

I  think something like this will gives us some results:

**`Renderer.scala`**
```scala

  implicit object dfAsWidget extends Renderer[DataFrame] {
    def render(x: DataFrame) = x match {
      case x if x.take(1).isEmpty => widgets.layout(0, Seq(widgets.text("")))
      case _ => display(x.take()
    }
  }
```


** `package.scala`**

**Comment by EronWright** [comment]
Incremental progress sounds good.   But please consider, a DataFrame does not contain Scala types, it contains SQL types.    The intended way to reason about the schema is not to use reflection but to use the schema property.   The best results will come from leveraging the provided schema and column information.

Given that JSON is involved in rendering, it would probably be worthwhile to look at the JSON conversion support in DataFrame itself (start [here](https://spark.apache.org/docs/1.3.0/sql-programming-guide.html#json-datasets)).  Granted it converts from JSON->DataFrame (the opposite of what we need here), but for symmetry's sake an examination is worthwhile.

**Comment by andypetrella** [comment]
Mmmh indeed, so we might have more stuffs at our disposal to do something tightly coupled.

Now, only "sequences" are taken into account, so if we stick with that, we'll lose the DataFrame, schema and so on, because we'll end up with `Row`s :-S.

So we really need something dedicated to `DataFrame`, that would include maybe also a tab for the schema and one for the plan.

That leads to the conclusion that we need a dedicated `display` function in `widgets` that will create the things we can plot or simply render the content. And based on your helpful pointers, we won't need the so-called `MagicRenderPoint` anymore.

At first very first step we can try this out then, but simply drawing a table, because the second step will refactor the `Chart`s to work based on a simple Json representation (or something like that)

**Comment by andypetrella** [comment]
This https://github.com/andypetrella/spark-notebook/pull/214 has been merged.
You can check `sql/DataFrame.snb` for some examples.

Of course, this can be extended horizontally (for instance a tab for the schema) and vertically (more charts, more customization)

@EronWright if you can give it a try, we can see what would be absolutely necessary to complete this first iteration on this?

**Comment by EronWright** [comment]
Andy I will have a look tonight.

Sent from my phone.
________________________________

**Comment by andypetrella** [comment]
cool :-)

**Comment by andypetrella** [comment]
Based on our discussions it'd be okay to close this one right? And directly create some enhancement tickets behind :-D

wdyt?

**Comment by EronWright** [comment]
Closing, to open new enhancement issues.

**Comment by EronWright** [comment]
Note that the improvements discussed above were implemented in #246.
